python/ray/data/_internal/pandas_block.py

```diff
@@ -300,11 +425,94 @@ def aggregate_combined_blocks(
         aggs: Tuple[AggregateFn],
         finalize: bool,
     ) -> Tuple["pandas.DataFrame", BlockMetadata]:
-        # TODO (kfstorm): A workaround to pass tests. Not efficient.
-        block, metadata = ArrowBlockAccessor.aggregate_combined_blocks(
-            [BlockAccessor.for_block(block).to_arrow() for block in blocks],
-            key,
-            aggs,
-            finalize,
+        """Aggregate sorted, partially combined blocks with the same key range.
+
+        This assumes blocks are already sorted by key in ascending order,
+        so we can do merge sort to get all the rows with the same key.
+
+        Args:
+            blocks: A list of partially combined and sorted blocks.
+            key: The column name of key or None for global aggregation.
+            aggs: The aggregations to do.
+            finalize: Whether to finalize the aggregation. This is used as an
+                optimization for cases where we repeatedly combine partially
+                aggregated groups.
+
+        Returns:
+            A block of [k, v_1, ..., v_n] columns and its metadata where k is
+            the groupby key and v_i is the corresponding aggregation result for
+            the ith given aggregation.
+            If key is None then the k column is omitted.
+        """
+
+        stats = BlockExecStats.builder()
+        key_fn = (lambda r: r[r._row.columns[0]]) if key is not None else (lambda r: 0)
+
+        iter = heapq.merge(
+            *[PandasBlockAccessor(block).iter_rows() for block in blocks], key=key_fn
+        )
+        next_row = None
+        builder = PandasBlockBuilder()
+        while True:
```

Comment:
nit: this merge-aggregate loop is also mostly same as Arrow's function. Non-blocking comment as well.